### PR TITLE
Ex4

### PR DESCRIPTION
**Pull request by ChipLuxury-EWA**
THANKS.
knowing bugs:
1 - fetching the same pokemon id works and it should not.
2 - 'clear all' button not working, I need to add delete req to new delete/all route...

### The requirements:

- [x] Create your express backend (include separate `dist` and `server` folders)
- [x] Your `server.js` file should have all the express boilerplate and host your `dist` directory to any client that requests it (hint: you'll need to `.use` the `express.static` method)
- [x] Create an `api.js` file that acts as the 'controller' of your backend, handling all the routes (endpoints)
- [x] Create separate endpoints for (1) fetching all the todo items, (2) creating a new one, and (3) deleting an existing one (hint: don't forget `bodyParser`)
- [x] Move the pokemon fetching code to the backend - use `axios` instead of `fetch` for your requests
- [x] On app load (i.e. when a user enters the page) it should fetch all the todo items and render them

In terms of the front end, it will look the same as before:

But now when you refresh the page **the data should still be there**

### Bonus

- [x] Create a [middleware](https://expressjs.com/en/guide/using-middleware.html) that makes a log each time a user accesses any of the routes (you can just do a `console.log`)
- [x] Handle server errors elegantly. Specifically, if anything goes wrong the user should see an error message (ideally, not an alert) with an explanation of what went wrong instead of crashing the page
- [ ] Add a loder/spinner to the page that indicates the client is waiting for an async operation (e.g. a call to the server) to finish
- [ ] Add simple caching to your server. If a user requests for the same pokemon ID three times in the same minute, for example, it should only make a request to the Pokemon API once. You can use a simple in-memory data structure for your cache

video:

https://user-images.githubusercontent.com/53507364/174037437-9cb0158d-d448-4d71-85fb-4be0eb164909.mp4

### CODE REVIEW

File: src/ex4/apiTest.http [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# @name healthCheck
```
Comment (barcohen2) [inline src/ex4/apiTest.http:1]: not sure what this file for?

File: src/ex4/tasks/task.txt [hunk 2]
```diff
@@ -0,0 +1,4 @@
+asdf
```
Comment (barcohen2) [inline src/ex4/tasks/task.txt:1]: not relevant file for the PR

File: src/ex4/dist/itemManger.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import { getAllTasks, deleteTask,addTask } from "./clients/item_client.js";
```
Comment (barcohen2) [inline src/ex4/dist/itemManger.js:1]: Cool! this way we keep ItemManager to manage the items and item_client is responsible for fetching items. well done!

File: src/ex4/dist/app.js [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+        const newButton = document.createElement("button");
+        //i tag is for font awesome
+        newButton.innerHTML = `<i class="fas fa-${buttonIcon}" ></i>`;
+        newButton.classList.add(buttonClass);
+        return newButton;
+    };
+
+    createPokemonSprit = async (pokemonName) => {
+        const newImg = document.createElement("img");
+        newImg.src = await pokemonClient.getPokemonImgByName(pokemonName);
```
Comment (barcohen2) [inline src/ex4/dist/app.js:85]: pokemon client logic should be used only in server. you shouldn't have pokemon client on both ends.

File: src/ex4/dist/app.js [hunk 5]
```diff
@@ -0,0 +1,220 @@
+import { ItemManager } from "./itemManger.js";
+import { PokemonClient } from "./clients/pokemonClient.js";
```
Comment (barcohen2) [inline src/ex4/dist/app.js:2]: shouldn't use pokemonclient in the client side

File: src/ex4/server/clients/pokemon_client.js [hunk 6]
```diff
@@ -0,0 +1 @@
+// The Pokemon Client (using axios) goes here
```
Comment (barcohen2) [inline src/ex4/server/clients/pokemon_client.js:1]: why didn't you implement pokemonClient in this folder? please don't keep empty files in your PRs

File: src/ex4/dist/clients/pokemonClient.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
+const URL = "https://pokeapi.co/api/v2/pokemon";
```
Comment (barcohen2) [inline src/ex4/dist/clients/pokemonClient.js:1]: pokemon client logic should exists only in the server

File: src/ex4/server/routes/v1/task.routes.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
+import express from "express";
+import {
+    readAndPrintAllTodos,
+    handleInput,
+    deleteTodo,
+} from "../../services/item_manager.js";
+//this route file import controller stuff
+const router = express.Router();
+
+router
```
Comment (barcohen2) [inline src/ex4/server/routes/v1/task.routes.js:10]: image
I recommend using this syntax, as it more readable and allow you to handle the response if you need to

File: src/ex4/server/server.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
+// Express boilerplate, hosting the `dist` file, connecting to the routes
```
Comment (barcohen2) [inline src/ex4/server/server.js:1]: app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: false }));

why didnt you user bodyParser?

File: src/ex4/server/services/item_manager.js [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+export const setFolderAndFile = async (folder, file) => {
+    folderAndFileInit(folder, file);
+};
+
+const addNewTask = (newTask, res) => {
+    console.log(chalk.bgCyan(`Adding ${newTask} to task list`));
+    addNewLine(process.env.FOLDER_NAME, process.env.FILE_NAME, newTask);
+    return true;
+};
+// export const addNewTask = (newTask) => {
```
Comment (barcohen2) [inline src/ex4/server/services/item_manager.js:24]: dont leave comments in the code

File: src/ex4/apiTest.http (hunk 1 above)
Comment (ChipLuxury-EWA) [reply src/ex4/apiTest.http:1]: Good morning this file is part of vsc addon named 'rest client'. This addon lets you send and receive http requests inside vsc, (even save jwt auth in header bearer)

File: src/ex4/tasks/task.txt (hunk 2 above)
Comment (ChipLuxury-EWA) [reply src/ex4/tasks/task.txt:1]: Do i have options to not include file in PR or i need to add this file to .gitignore?

File: src/ex4/server/server.js (hunk 9 above)
Comment (ChipLuxury-EWA) [reply src/ex4/server/server.js:1]: I used express built-in body parser as you can see in line 21.
That way I don't need to install another (body-parser) package...

File: src/ex4/server/server.js [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+dotenv.config();
+setFolderAndFile("tasks", "task");
+
+const app = express();
+app.use(morgan("common"))
+
+const __filename = fileURLToPath(import.meta.url);
+const __dirname = path.dirname(__filename);
+
+app.use(express.json());
```
Comment (ChipLuxury-EWA) [inline src/ex4/server/server.js:21]: this line replace:
```
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: false }));
```

File: src/ex4/server/clients/pokemon_client.js (hunk 6 above)
Comment (ChipLuxury-EWA) [reply src/ex4/server/clients/pokemon_client.js:1]: I will replace the file location

File: src/ex4/dist/clients/pokemonClient.js (hunk 7 above)
Comment (ChipLuxury-EWA) [reply src/ex4/dist/clients/pokemonClient.js:1]: i'm useing it to render the pokemon pic in the frontend, is there any tip for best-practice?

File: src/ex4/dist/app.js (hunk 4 above)
Comment (ChipLuxury-EWA) [reply src/ex4/dist/app.js:85]: in the frontend it used to get the pokemon image in order to render it.

File: src/ex4/dist/clients/pokemonClient.js (hunk 7 above)
Comment (barcohen2) [reply src/ex4/dist/clients/pokemonClient.js:1]: return the image url from server when fetching the pokemon

**Pull request merged**